src/resource_manager/resource_manager.py

```diff
@@ -0,0 +1,250 @@
+
+    Args:
+        resource (AnyResource): The kubernetes resource.
+
+    Returns:
+        typing.Optional[str]: The resource name, or None if not set.
+    """
+    if resource is None:
+        return None
+    if resource.metadata is None:
+        return None
+    if resource.metadata.name is None:
+        return None
+    return resource.metadata.name
+
+
+class ResourceManager(typing.Protocol[AnyResource]):
+    """Abstract base class for a generic Kubernetes resource controller."""
+
+    @property
+    @abc.abstractmethod
+    def _name(self) -> str:
+        """Abstract property that returns the name of the resource type.
+
+        Returns:
+            Name of the resource type.
+        """
+
+    @property
+    @abc.abstractmethod
+    def _namespace(self) -> str:
+        """Abstract property that returns the namespace of the controller.
+
+        Returns:
+            The namespace.
+        """
+
+    @abc.abstractmethod
+    def _gen_resource_from_definition(self, definition: ResourceDefinition) -> AnyResource:
+        """Abstract method to generate a resource from ingress definition.
+
+        Args:
+            definition: Ingress definition to use for generating the resource.
+        """
+
+    @abc.abstractmethod
+    def _create_resource(self, resource: AnyResource) -> None:
+        """Abstract method to create a new resource in a given namespace.
+
+        Args:
+            resource: The resource object to create.
+        """
+
+    @abc.abstractmethod
+    def _patch_resource(self, name: str, resource: AnyResource) -> None:
+        """Abstract method to patch an existing resource in a given namespace.
+
+        Args:
+            name: The name of the resource to patch.
+            resource: The modified resource object.
+        """
+
+    @abc.abstractmethod
+    def _list_resource(self) -> typing.List[AnyResource]:
+        """Abstract method to list resources in a given namespace based on a label selector."""
+
+    @abc.abstractmethod
+    def _delete_resource(self, name: str) -> None:
+        """Abstract method to delete a resource from a given namespace.
+
+        Args:
+            name: The name of the resource to delete.
+        """
+
+    def define_resource(
+        self,
+        definition: ResourceDefinition,
+    ) -> AnyResource:
+        """Create or update a resource in kubernetes.
+
+        Args:
+            definition: The ingress definition
+
+        Returns:
+            The name of the created or modified resource.
+
+        Raises:
+            InvalidResourceError: If the generated resource is invalid.
+        """
+        resource_list = self._list_resource()
+        resource = self._gen_resource_from_definition(definition)
+        res_name = resource_name(resource)
+        if not res_name:
+            raise InvalidResourceError("Missing resource name.")
+
+        resources = [resource_name(r) for r in resource_list if resource_name(r) is not None]
```

Comment:
Put the related statements close together (`resource_list` and `resource` definition)
And the names are a bit ambiguous, try to be more specific (`new_resource` or `existing_resources`)